keps/sig-apps/0003-kubernetes-application-api.md

```diff
@@ -0,0 +1,1120 @@
+    Spec ApplicationSpec `json:"spec,omitempty" protobuf:"bytes,2,opt,name=spec"`
+    
+    // Status contains the Status of the application with respect to the installed components.
+    Status ApplicationStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
+}
+```
+
+The purpose of the Application Kind is to aggregate the components and dependencies of a Kubernetes application such 
+that they are explicit and discoverable. An Application's `Spec.Components` is a mapping of the components in an 
+application to the GVK (Group Version Kind) used to realize the component. Each component must be labeled such that its 
+owning Application's `Spec.Selector` correctly matches the component's labels. Clients of the API Server can use the GVK 
+information, along with the label selector, to list or watch the components in an Application. Additionally, each 
+component should be annotated with a reference to the Application.
+
+An Application's `Spec.Dependencies` is a list of other Applications on which the application depends. Clients can use 
+this field to discover and retrieve other Applications that the containing Application requires to be present in the 
+system.
+
+While we attempt to represent many properties of the Application object with explicit fields, we realize that the 
+current list is not exhaustive, and we do not endeavor to develop an exhaustive list prior to releasing the 
+Application Kind. We expect users to use the fields of ObjectMeta (i.e. Labels and Annotations) for custom use cases. 
+When the community forms a consensus that a particular label or annotation has become prevalent across a large 
+number of tools, we will promote that metadata to a field. Note that this can be done in a backward compatible manner 
+without respect to the stability of the Application Kind.
+
+#### Validation
+Upon Application creation, The API Server shall perform the following validation.
+
+1. The API Server shall validate that the Application is a valid ObjectMeta with a name that is a valid DNS subdomain. 
+In effect, the API Server will simply validate that the Application is a valid Kubernetes resource.
+1. The API Server will validate that the Application's `Spec.Selector` is a valid, non-nil, set-based Selector.
+1. If the Application has a `.Spec.HealthCheck`, the API Server will validate that the `Template` of the 
+ApplicationHealthCheck is a valid Pod with a readiness check.
+
+The API Server shall perform the following validation upon Application mutation.
+
+1. The API server shall ensure that the new version of the Application is valid using the above definition of a valid 
+newly created Application instance. 
+1. The API server shall ensure that the Applications `Spec.Selector` has not been mutated. This latter condition has 
+been implemented in many places in the Kubernetes API to prevent unintentional orphaning. This is our purpose here.
+
+#### Generation Incrementation
```

Comment:
It took a while to realize that you are using the term "generation" to mean version or revision. I thought it was responsible for generating something, which is the way you use the term elsewhere in this doc. Maybe change to revision, which is a less ambiguous term?

Reply:
This is the existing metadata.generation field in Kubernetes.
https://github.com/kubernetes/api/blob/master/core/v1/types.go#L140
https://github.com/kubernetes/kubernetes/issues/7328

Reply:
ObservedGeneration is the standard. With the work in flight on CRD subresources, this will be the only way to achieve the monotonically increasing register that we use in core resources.

Reply:
If we add a manually managed AppVersion as suggested by prydonius , then I think there is less issue with the name ObservedGeneration.